tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,228 @@
+    /// Path to root.json for this repo
+    root_role_path: PathBuf,
+
+    #[structopt(default_value = "100", long)]
+    /// Randomly samples and retrieves specified percentage of targets
+    percent_target_files: u8,
+}
+
+fn attempt_read_target<T>(
+    repo: &Repository<'_, T>,
+    target: String,
+) -> Result<Option<impl Read>, repo_error::Error>
+where
+    T: tough::Transport,
+{
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        match repo.read_target(&target) {
+            Err(err) => match err {
+                // Retry if it's a transport failure, might be due to intermittent connection issues
+                tough::error::Error::Transport { .. } => {
+                    error!(
+                        "Repo target fetch failure on attempt {} out of {}",
+                        attempt, max_attempts
+                    );
+                    if attempt >= max_attempts {
+                        return Err(repo_error::Error::ReadTarget {
+                            target,
+                            source: err,
+                        });
+                    }
+                }
+                // For other errors we should just return immediately and not retry
+                err => {
+                    return Err(repo_error::Error::ReadTarget {
+                        target,
+                        source: err,
+                    })
+                }
+            },
+            Ok(target_reader) => return Ok(target_reader),
+        };
+    }
+
+    Ok(None)
+}
+
+fn attempt_download_target(mut reader: impl Read, target: String) -> Result<(), Error> {
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        info!("Downloading target: {}", target);
+        if let Err(err) = io::copy(&mut reader, &mut io::sink()) {
+            error!(
+                "Download error on attempt {} out of {}: {}",
+                attempt, max_attempts, err
+            );
+            if attempt >= max_attempts {
+                return Err(error::Error::TargetDownload {
+                    target,
+                    source: err,
+                });
+            }
+            continue;
+        }
+        // Successfully downloaded target
+        break;
+    }
+
+    Ok(())
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them.
+fn retrieve_percentage_of_targets<T>(repo: &Repository<'_, T>, percentage: u8) -> Result<(), Error>
+where
+    T: tough::Transport,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+    let mut sampled_targets: Vec<String> = targets.keys().map(|key| key.to_string()).collect();
+    sampled_targets = sampled_targets
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+        .collect();
```

Comment:
I'm not sure we need to clone twice here; I was able to get this to work with a `HashMap<String, foo>` like we get from `targets().signed.targets`:
```
    let keys = map.keys().collect::<Vec<&String>>();
    for selected in keys.choose_multiple(&mut rng, 1) {
```
Would that work here?  We're calling `to_string` / `clone` below anyway, where we need ownership...

Reply:
After switching to using async, the compiler keeps complaining about the lifetime of `targets` being not long enough. So I ended up still having to clone the keys and the sampled targets.

Reply:
I think it might just need the `.choose_multiple.cloned`, and not the prior `String`/`cloned` when defining `targets` because each `target` is all that goes into the async block.  (Alternatively, maybe `let target_str = target.to_string();` before the `async move`, and using `target_str` inside the `async move`, would be enough to clarify that only a single owned value is moving in.)

This is all just a performance nit, we can move on if it's being difficult!